### PR TITLE
Documenting new algorithm Detect Dataset Changes

### LINKED ISSUE #4788: [FEATURE][processing] Add new algorithm "Detect Dataset Changes"
Original commit: https://github.com/qgis/QGIS/commit/8c8d48bb8fe82ef8654dd65233e3265f4c3e524e by nyalldawson

This algorithm compares two vector layers, and determines which features
are unchanged, added or deleted between the two. It is designed for comparing
two different versions of the same dataset.

When comparing features, the original and revised feature geometries will be
compared against each other. Depending on the Geometry Comparison Behavior setting,
the comparison will either be made using an exact comparison (where geometries must
be an exact match for each other, including the order and count of vertices) or a
topological comparison only (where are geometries area considered equal if all of
the their component edges overlap. E.g. lines with the same vertex locations but
opposite direction will be considered equal by this method). If the topological
comparison is selected then any z or m values present in the geometries will not
be compared.

By default, the algorithm compares all attributes from the original and revised
features. If the Attributes to Consider for Match parameter is changed, then only
the selected attributes will be compared (e.g. allowing users to ignore a timestamp
or ID field which is expected to change between the revisions).

If any features in the original or revised layers do not have an associated geometry,
then care must be taken to ensure that these features have a unique set of
attributes selected for comparison. If this condition is not met, warnings will be
raised and the resultant outputs may be misleading.

The algorithm outputs three layers, one containing all features which are considered
to be unchanged between the revisions, one containing features deleted from the
original layer which are not present in the revised layer, and one containing features
add to the revised layer which are not present in the original layer.

### PR DESCRIPTION
**Pull request by rosaguilar**
Goal:

Ticket(s): Fix #4788

- [ ] Backport to LTR documentation is required

### CODE REVIEW

**Comment by m-kuhn** [comment]
Wow, nice work. I love the image to explain the alg.

It looks like the features tagged "deleted" are still there in set 2. Is this an error in my understanding, the alg or the image?

**Comment by rosaguilar** [comment]
> Wow, nice work. I love the image to explain the alg.
> 
> It looks like the features tagged "deleted" are still there in set 2. Is this an error in my understanding, the alg or the image?

Thanks for catching this. 

> Wow, nice work. I love the image to explain the alg.
> 
> It looks like the features tagged "deleted" are still there in set 2. Is this an error in my understanding, the alg or the image?

Thanks for catching this. A new image was uploaded.

**Comment by rosaguilar** [comment]
> Looks fine to me.

@havatv, Can you please check (in the code) the names of "unchanged", "deleted", and "added".
I assumed those names but I was waiting for your feedback. I already sent a mail to Nyall but not reply yet.

**Comment by havatv** [comment]
> @havatv, Can you please check (in the code) the names of "unchanged", "deleted", and "added".
> I assumed those names but I was waiting for your feedback. I already sent a mail to Nyall but not reply yet.

Both the labels and names are correct.
However, the three output parameters that give the count for each group should also be included:

* "UNCHAGED_COUNT" ( "Count of unchanged features" )
* "ADDED_COUNT" ( "Count of features added in revised layer" )
* "DELETED_COUNT" ("Count of features deleted from original layer")

 (@nyalldawson, note the spelling mistake in the code: "UNCHAGED" -> "UNCHANGED"):

**Comment by nyalldawson** [comment]
>  I already sent a mail to Nyall but not reply yet.

Sorry, don't recall receiving that email

>  note the spelling mistake in the code: "UNCHAGED" -> "UNCHANGED"

oops will fix, thanks!

**Comment by nyalldawson** [comment]
@rosaguilar 
> Can you please check (in the code) the names of "unchanged", "deleted", and "added".
I assumed those names but I was waiting for your feedback. I already sent a mail to Nyall but not reply yet.

sorry, found that email. It's a QGIS bug, I'm fixing now

**Comment by rosaguilar** [comment]
Thanks @havatv and @nyalldawson. I added the missing output parameters (log), not sure of the labels. I also replaced the figure to delete the caption (I think it is not needed).

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 1]
```diff
@@ -537,7 +537,115 @@ Outputs
...
+     - [same as input]
+     - Specify the output vector layer containing the deleted features. One of:
+
+       * Create Temporary Layer (``TEMPORARY_OUTPUT``)
+       * Save to File...
+       * Save to Geopackage...
+       * Save to PostGIS Table...
+
+       The file encoding can also be changed here.
+   * - **Unchanged count**
```
Comment (havatv) [inline docs/user_manual/processing_algs/qgis/vectorgeneral.rst]:
```suggestion
   * - **Count of unchanged features**
```

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 2]
```diff
@@ -537,7 +537,115 @@ Outputs
...
+     - ``UNCHANGED_COUNT``
+     - [number]
+     - Count of unchanged features.
+   * - **Added count**
```
Comment (havatv) [inline docs/user_manual/processing_algs/qgis/vectorgeneral.rst]:
```suggestion
   * - **Count of features added in revised layer**
```

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 3]
```diff
@@ -537,7 +537,115 @@ Outputs
...
+     - ``ADDED_COUNT``
+     - [number]
+     - Count of features added in revised layer.
+   * - **Deleted count**
```
Comment (havatv) [inline docs/user_manual/processing_algs/qgis/vectorgeneral.rst]:
```suggestion
   * - **Count of features deleted from original layer**
```

**Comment by havatv** [comment]
> Thanks @havatv and @nyalldawson. I added the missing output parameters (log), not sure of the labels. I also replaced the figure to delete the caption (I think it is not needed).

It was actually the lables that were placed in parenthesis after the names. See my suggestions.

**Comment by rosaguilar** [comment]
> > Thanks @havatv and @nyalldawson. I added the missing output parameters (log), not sure of the labels. I also replaced the figure to delete the caption (I think it is not needed).
> 
> It was actually the lables that were placed in parenthesis after the names. See my suggestions.

Thanks, suggestions applied.

**Comment by rosaguilar** [comment]
> > Thanks @havatv and @nyalldawson. I added the missing output parameters (log), not sure of the labels. I also replaced the figure to delete the caption (I think it is not needed).
> 
> It was actually the lables that were placed in parenthesis after the names. See my suggestions.

Sorry. I thought that it was the description :)

**Comment by havatv** [comment]
> Sorry. I thought that it was the description :)

No problem! Thanks for your efforts.

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 4]
```diff
@@ -537,7 +537,115 @@ Outputs
...
+.. list-table::
+   :header-rows: 1
+   :widths: 20 20 20 40
+   :stub-columns: 0
+
+   * - Label
+     - Name
+     - Type
+     - Description
+   * - **Unchanged features**
```
Comment (DelazJ) [inline docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: @havatv Are these outputs shown in the dialog no longer part of the parameters (eg https://docs.qgis.org/3.10/en/docs/user_manual/processing_algs/qgis/vectorcreation.html#import-geotagged-photos)?

Comment (havatv) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: @rosaguilar, the three output datasets/layers must also be described among the parameters (they should be present both in the Parameters and Output tables) - see for instance @DelazJ's reference for how it is normally done.

Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: Thanks, @DelazJ, and @havatv.

Comment (havatv) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: The *Type* of the output should be [same as input] (not [vector: any]) in both the Parameters and Output tables.

Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: ok. but what is **input**?
I followed the example (https://docs.qgis.org/3.10/en/docs/user_manual/processing_algs/qgis/vectorcreation.html#import-geotagged-photos).  or 
https://docs.qgis.org/3.10/en/docs/user_manual/processing_algs/qgis/vectorcreation.html#id34
I was thinking of [same as Original], as "Original" is the name of the "Original layer".
Its confusing

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 5]
```diff
@@ -537,7 +537,127 @@ Outputs
...
+     - [enumeration]
+       
+       Default: 1
+     - Defines the criteria for comparison. Options:
+       
+       * 0 --- Exact Match: includes the order and vertices count of geometries
+       * 1 --- Tolerant Match (Topological Equality): geometries are considered equal
+   * - **Unchanged features**
+     - ``UNCHANGED``
+     - [vector: any]
```
Comment (m-kuhn) [inline docs/user_manual/processing_algs/qgis/vectorgeneral.rst]:
```suggestion
     - [vector: same as Original layer]
```

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst (hunk 4 above)
Comment (m-kuhn) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: I've left a suggestion the way I think it should be

Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: Thanks @m-kuhn. Suggestion added.

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 6]
```diff
@@ -537,7 +537,127 @@ Outputs
...
+
+       * Create Temporary Layer (``TEMPORARY_OUTPUT``)
+       * Save to File...
+       * Save to Geopackage...
+       * Save to PostGIS Table...
+
+       The file encoding can also be changed here.
+   * - **Deleted features**
+     - ``DELETED``
+     - [vector: any]
```
[the remark this answers is not in the document]
Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: Thanks @m-kuhn,  Same here.

File: docs/user_manual/processing_algs/qgis/vectorgeneral.rst [hunk 7]
```diff
@@ -537,7 +537,127 @@ Outputs
...
+
+       * Create Temporary Layer (``TEMPORARY_OUTPUT``)
+       * Save to File...
+       * Save to Geopackage...
+       * Save to PostGIS Table...
+
+       The file encoding can also be changed here.
+   * - **Added features**
+     - ``ADDED``
+     - [vector: any]
```
[the remark this answers is not in the document]
Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: Thanks @m-kuhn,  this was already done in my previous commit.

[the remark this answers is not in the document]
Comment (m-kuhn) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: I don't think so. Click on the "Commit suggestion" button -- or even better use "Add suggestion to batch" on the files tab to create a new commit with the 3 remaining suggestions.

[the remark this answers is not in the document]
Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: > I don't think so. Click on the "Commit suggestion" button -- or even better use "Add suggestion to batch" on the files tab to create a new commit with the 3 remaining suggestions.

@m-kuhn, sorry for the questions, I am trying to understand how this works. Commiting suggestions is te only way to address changes? I edit directly the file, after your first suggestion, and I changed the three ocurrences

[the remark this answers is not in the document]
Comment (m-kuhn) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: You can also modify it yourself. But you need to change 6 occurrences (3 x parameters, 3 x outputs)

[the remark this answers is not in the document]
Comment (rosaguilar) [reply docs/user_manual/processing_algs/qgis/vectorgeneral.rst]: @m-kuhn, thanks. Hopefully, it is done now.

**Comment by m-kuhn** [comment]
Awesome work @rosaguilar :rocket:

**Comment by rosaguilar** [comment]
> Awesome work @rosaguilar 🚀

Thanks, @m-kuhn for reviewing and for the encouraging words.

**Comment by m-kuhn** [comment]
@havatv @DelazJ is there anything missing in here from your perspective?
For me, it's good to ship :ship:

**Comment by DelazJ** [comment]
Thanks @rosaguilar Great work as usual.

**Comment by rosaguilar** [comment]
Thanks to all the reviewers too.

**Pull request merged**